File: 53-maximum-subarray/53-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        dp = nums[0]
        maxSub = dp
        for i in range(1, len(nums)):
            dp = max(nums[i] + dp, nums[i])
            maxSub = max(maxSub, dp)
        return maxSub

# time and space complexity
# time: O(n)
# space: O(1)

File: 53-maximum-subarray/53-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        for (int i = 1; i < nums.size(); i++){
            nums[i] = max(nums[i] + nums[i - 1], nums[i]);
        }
        int m = nums[0];
        for (int i = 1; i < nums.size(); i++){
            m = max(m, nums[i]);
        } 
        return m;
    }
};

// # time and space complexity
// # time: O(n)
// # space: O(1)